From 71f99065c4b68835bbb32eb84898850037a16c58 Mon Sep 17 00:00:00 2001
Message-Id: <71f99065c4b68835bbb32eb84898850037a16c58.1413836944.git.chang-joon.lee@intel.com>
In-Reply-To: <5d8aa9994fff43965a6fab8c9a528b8b47a9d624.1413836944.git.chang-joon.lee@intel.com>
References: <5d8aa9994fff43965a6fab8c9a528b8b47a9d624.1413836944.git.chang-joon.lee@intel.com>
From: Sourab Gupta <sourab.gupta@intel.com>
Date: Wed, 1 Oct 2014 20:35:56 +0530
Subject: [PATCH 64/71] FOR_UPSTREAM [VPG]: drm/i915/vlv: Fix the issue of
 artifacts on use of stolen mem

This commit fixes a bug introduced by a previous commit
5cd611028df02f9d5361719791751648ca987648

This was leading to artifacts seen on display if user framebuffer was allocated
from stolen memory instead of shmem. These artifacts were introduced only in a
specific usecase of using mmap_gtt to access the buffers and then giving them to
render engine.
More specifically, while pinning the object(before giving to gpu), the fake mmap
offset of the object(acquired during previous mmap_gtt) was erroneously
released.The same mmap offset could be given to another object (while first
object is still in use).
The 'fake offset' has to be reserved for the entire life span of the object, because
if not done so, then the 'i915_gem_release_mmap' function may not work properly,
whenever there is a fence stealing, because then fence register will be
reconfigured as per some other object's properties, but the Page tables
entries corresponding to 'mmap_gtt' mapping will not get cleared out.
This will cause an incorrect access to the buffer.
This patch provides a fix for this issue.

Issue: GMINL-1730
Change-Id: I4715463cb3f5ea3d51f8328eba23ff26e78d331a
Signed-off-by: Sourab Gupta <sourab.gupta@intel.com>
Signed-off-by: Akash Goel <akash.goel@intel.com>
---
 drivers/gpu/drm/i915/i915_gem_stolen.c |    4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/i915/i915_gem_stolen.c b/drivers/gpu/drm/i915/i915_gem_stolen.c
index 94ef905..2ba4f4c 100644
--- a/drivers/gpu/drm/i915/i915_gem_stolen.c
+++ b/drivers/gpu/drm/i915/i915_gem_stolen.c
@@ -549,7 +549,9 @@ i915_gem_object_move_to_stolen(struct drm_i915_gem_object *obj)
 
 	/* Set up the object to use the stolen memory,
 	 * backing store no longer managed by shmem layer */
-	drm_gem_object_release(&(obj->base));
+	if (obj->base.filp)
+		fput(obj->base.filp);
+
 	obj->base.filp = NULL;
 	obj->ops = &i915_gem_object_stolen_ops;
 
-- 
1.7.9.5

